### PR TITLE
Feature/recreate previous encryption

### PR DESCRIPTION
**Pull request by emm1R**
This PR will make it possible to recreate a previous encryption. This is needed if one wishes to stream the encrypted results but also calculate a checksum for the file. The data key, and header and segment nonces prevented this from happening so these are now exposed through Crypt4GHWriter.Rands. To recreate the same encryption, these values need to be passed to NewCrypt4GHWriterWithRands().

### CODE REVIEW

File: streaming/out.go [hunk 1]
```diff
@@ -20,22 +20,41 @@ type Crypt4GHWriter struct {
...
+
+type WriterRands struct {
+	dataKey      [chacha20poly1305.KeySize]byte
+	headerNonces []*[chacha20poly1305.NonceSize]byte
+	bodyNonces   []*[chacha20poly1305.NonceSize]byte
 }
 
 // NewCrypt4GHWriter method constructs streaming.Crypt4GHWriter instance from io.Writer and corresponding keys.
-func NewCrypt4GHWriter(writer io.Writer, writerPrivateKey [chacha20poly1305.KeySize]byte, readerPublicKeyList [][chacha20poly1305.KeySize]byte, dataEditList *headers.DataEditListHeaderPacket) (*Crypt4GHWriter, error) {
+func NewCrypt4GHWriter(
```
Comment (pontus) [inline streaming/out.go:33]: Could we keep the old interface please (and have this as e.g. `NewCrypt4GHWriterWithRands` or similar)?

Comment (emm1R) [reply streaming/out.go:33]: ok

File: model/headers/headers.go [hunk 2]
```diff
@@ -214,10 +215,15 @@ func (h Header) MarshalBinary() (data []byte, err error) {
...
 	for _, headerPacket := range h.HeaderPackets {
+		if h.Nonces != nil {
+			headerPacket.Nonce = h.Nonces[0]
+			h.Nonces = h.Nonces[1:]
+		}
 		marshalledHeaderPacket, err := headerPacket.MarshalBinary()
 		if err != nil {
 			return nil, err
 		}
+		h.Nonces = append(h.Nonces, headerPacket.Nonce)
```
Comment (pontus) [inline model/headers/headers.go:226]: Maybe copy `h.Nonces` to a local variable and operate on that instead of removing and adding back?

**Comment by emm1R** [comment]
> (I'm not sure I think it's a good idea with the `Nonces` in the header rather than the `headerPacket`. I guess it makes things simpler, but it makes me confused about which one is considered at which point.

I can change that

File: model/headers/headers.go [hunk 3]
```diff
@@ -213,8 +213,8 @@ func (h Header) MarshalBinary() (data []byte, err error) {
 	if err != nil {
 		return nil, err
 	}
-	for _, headerPacket := range h.HeaderPackets {
-		marshalledHeaderPacket, err := headerPacket.MarshalBinary()
+	for i := range h.HeaderPackets {
```
Comment (pontus) [inline model/headers/headers.go:216]: Not wrong, but this change does not seem needed if no longer keeping the nonces in the header.

**Comment by viklund** [comment]
I don't understand the use case here. It should be possible to calculate the checksum at the same time as encrypting and streaming the file? The ingest service does this, it copies (by streaming) a file from inbox (an io.Reader) to internal storage (an io.Writer) while calculating the checksum?

https://github.com/neicnordic/sda-pipeline/blob/master/cmd/ingest/ingest.go#L244-L334

File: model/headers/headers.go (hunk 3 above)
Comment (emm1R) [reply model/headers/headers.go:216]: It is needed because otherwise the new Nonce is not reachable in other function.

**Comment by emm1R** [comment]
> I don't understand the use case here. It should be possible to calculate the checksum at the same time as encrypting and streaming the file? The ingest service does this, it copies (by streaming) a file from inbox (an io.Reader) to internal storage (an io.Writer) while calculating the checksum?
> 
> https://github.com/neicnordic/sda-pipeline/blob/master/cmd/ingest/ingest.go#L244-L334

In our case, we don’t want to save the entire encrypted file anywhere in memory. We are streaming it away with POST request. If the encryption was performed only once, the checksum could only be calculated after the entire file had been sent but we want to send it at the same time. That’s why we hoped we could perform the enryption twice and get the same result each time.

**Comment by pontus** [comment]
If that (checksum for sent file) is the only use case you have for now, I'm confident it can be done in a better way without these changes.

The way interfaces are used in golang means it's real easy to hook in an intermediate/passthrough Writer that does something (e.g. sends output to hash computation as well) without any more memory being used.

If it's still not clear how that could look, let me know and I'd be happy to cook up an example.

**Comment by pontus** [comment]
I took some minutes proactively and wrote up an example at https://github.com/pontus/writerdemo

**Pull request closed without merging**